### ISSUE TITLE
[Feature]: Support for MERGE syntax

### ISSUE THREAD

**Issue by kewinbrand** [babelfish-for-postgresql/babelfish_extensions#2920]
### What happened?

We are evaluating Babelfish for a long time now but are facing some basic problems due to the missing syntax "MERGE", it is heavily usesd by some CDC tools like DEBEZIUM. We also plan to use it with web development, where libraries like EF Core also rely on this syntax.
Is there any plan to support in the near future?

### Version

BABEL_4_X_DEV (Default)

### Extension

None

### Which flavor of Linux are you using when you see the bug?

Ubuntu (Default)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

**Comment by robverschoor** [comment]
T-SQL MERGE supports a number of commonly used syntactic and semantic aspects which PG does currently not support. Right now, the best workaround is to use the `-rewrite` feature of BabelfishCompass, but this is useful only when you control the actual SQL code (so with frameworks it may not be a solution).
For PG 17, it looks like some (not all) of these aspects will be addressed, so we should re-evaluate whether Babelfish can use that. 
In the mean time, could you please confirm whether your MERGE statements contain any of the following, which PG 17 does not seem to be supporting (as it looks currently):
- the target table is an updatable view
- the target table is a CTE
- use of recursive CTE in the USING clause is supported

**Comment by kewinbrand** [comment]
we do not use any of these statements you mentioned
for now we are considering pointing the CDC tool directly to the underlying babelfish postgres. we have tested the basics and it seems to have had no problems. is there any documentation or guide that explains the risks of doing this?

**Comment by scaliebe** [comment]
I have a working implementation of T-SQL MERGE for Babelfish and would like to contribute it, if PRs for this are welcome.

As anticipated [earlier in this thread](https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/2920#issuecomment-2340742224), the PG 17/18 features close the gap: the implementation maps T-SQL MERGE onto PostgreSQL's native MERGE (`WHEN NOT MATCHED BY SOURCE/TARGET`, `RETURNING` with `merge_action()`, `old`/`new` transition references), with the T-SQL syntax differences handled by query rewriting in the ANTLR layer, plus a small backend-parser addition and a one-line change in the engine (an executor-level guard that currently blocks `CMD_MERGE`).

Working so far, validated side-by-side against SQL Server 2022 with identical results:
- all `WHEN [NOT] MATCHED [BY SOURCE/TARGET]` variants with `AND` conditions, `INSERT`/`UPDATE`/`DELETE` actions, `DEFAULT VALUES`
- `OUTPUT` incl. `$action` (typed `nvarchar(10)`, named `$action`), `inserted.*`/`deleted.*`
- `OUTPUT ... INTO` (table variables and tables, rewritten as a data-modifying CTE)
- compound assignment in `SET`, table variables as target, MERGE in procedures, `@@ROWCOUNT`
- SQL Server error codes: Msg 8672 (row affected twice) and Msg 10713 (missing terminating semicolon)

Known deviations (documented): `MERGE TOP (n)` is rejected for now, the Msg 8672 message text is the PG wording, and `@@ROWCOUNT` after an OUTPUT clause returning rows to the client is 0 (matching existing Babelfish behavior for INSERT/UPDATE/DELETE with OUTPUT). Also covered by a new JDBC test file.

Since this spans both repos, I would open two linked PRs (same branch name, so CI picks up the engine branch). Before I do: is there anything you would want done differently, or internal work in progress I should be aware of?

**Comment by tanscorpio7** [comment]
@scaliebe Please go ahead with the PR, this is not being worked on internally.

A few suggestions:
1. **Develop this feature behind a superuser GUC (default off).** We only enable features after extensive validation, and having a feature flag avoids the need for reverts if issues are discovered.
2. **Please try to break the PR into smaller, incremental pieces.** For example: one PR to add the GUC and lift syntax errors, then a follow-up for the parser rewrite, and so on. This makes review easier and reduces risk.

**Comment by scaliebe** [comment]
Thanks — will do. The plan:

- **Engine PR**: the small executor change (lift the `CMD_MERGE` guard in `fireISTriggers()`), same branch name so CI picks it up.
- **Extensions PR 1**: add `babelfishpg_tsql.enable_tsql_merge` (superuser-only, default off) and basic MERGE support — backend parser + executor plumbing, canonical `MERGE INTO ... WHEN ... ;` syntax only, everything else still rejected. With the GUC off, behavior is unchanged.
- **Follow-up PRs**, each incremental and gated behind the GUC: the T-SQL syntax rewrites (omitted `INTO`, alias-qualified/compound `SET`, table hints, SQL Server error codes 8672/10713), then the `OUTPUT` clause incl. `OUTPUT ... INTO`.

Each PR comes with JDBC tests, including coverage that the GUC-off path keeps the current error.

**Comment by robverschoor** [comment]
It would be great if MERGE support could be added. Did you verify that the semantics of the various new clauses are identical to SQL Server? 

If the answer is yes, then the question is which aspects will remain unsupported (this is needed to update the Compass tool accordingly). I have not seen the following points discussed, so can you confirm whether the following are still correct?
- The target table cannot be an updatable view; also it cannot be an Materialized View
- A CTE for a table in the USING clause is supported, but not for recursive CTEs
- The target table can be a CTE in SQL Server, but not in PG (the same limitation exists for INSERT, UPDATE and DELETE)

As for the implementation, I would suggest to first build an extensive set of tests (see https://github.com/babelfish-for-postgresql/babelfish_extensions/tree/BABEL_6_X_DEV/test/JDBC/input) since this is a complex T-SQL feature with many possible variations (such as multiple WHEN clauses) all of which need to be covered by tests.

**Comment by scaliebe** [comment]
Yes — semantics were verified side-by-side against SQL Server 2022 with the same statements through the same client: all `WHEN` variants and their ordering, `OUTPUT` result rows and column naming/typing (`$action` as `nvarchar(10)`), `@@ROWCOUNT`, and the double-update error. Documented deviations so far: the Msg 8672 message *text* is the PG wording (number/severity match), `@@ROWCOUNT` after an `OUTPUT` clause returning rows to the client is 0 (matching existing Babelfish behavior for INSERT/UPDATE/DELETE with OUTPUT), and `MERGE TOP (n)` is rejected.

On your three points — I tested each on both sides; one is better than expected:
- **Updatable view as target: supported.** PG 17+ allows MERGE into auto-updatable views, and it works in Babelfish (verified; one hook needed a `CMD_MERGE` case, included in the PR). Not supported: views with T-SQL INSTEAD OF triggers (clean error), and materialized views (clean error; not reachable from plain T-SQL anyway).
- **CTE in the USING clause: supported; recursive CTE in USING: not supported** — PG raises `WITH RECURSIVE is not supported for MERGE statement` (SQL Server allows it). Compass-relevant.
- **CTE as target: not supported** (relation does not exist), same PG limitation as for INSERT/UPDATE/DELETE (SQL Server allows it). Compass-relevant.

So for Compass, the remaining unsupported aspects as of this PR series: `MERGE TOP (n)`, recursive CTE in `USING`, CTE as target, INSTEAD OF triggers on the target, and the two deviations above. `OUTPUT` incl. `OUTPUT ... INTO` will be supported (part 3 of the series).

On tests: agreed. The first PR (#4928) ships a JDBC test file covering the basics incl. the GUC-off path; the follow-ups extend it per feature (multiple/reordered `WHEN` clauses, `OUTPUT` variants, error cases). If there are specific scenarios Compass needs verified, let me know.

**Comment by tanscorpio7** [comment]
https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/4928
